File: 0446-arithmetic-slices-ii-subsequence/0446-arithmetic-slices-ii-subsequence.java
class Solution {
    public int numberOfArithmeticSlices(int[] nums) {
        int n = nums.length;
        int ans = 0;
        Map<Integer, Integer>[] f = new Map[n];
        for (int i = 0; i < n; ++i) {
            f[i] = new HashMap<Integer, Integer>();
            for (int j = 0; j < i; ++j) {
                long diff = (long) nums[i] - nums[j];
                if (diff <= Integer.MIN_VALUE || diff > Integer.MAX_VALUE) {
                    continue;
                }
                int d = (int) diff;
                int cnt = f[j].getOrDefault(d, 0);
                ans += cnt;
                f[i].put(d, f[i].getOrDefault(d, 0) + cnt + 1);
            }
        }
        return ans; 
    }
}

File: 0446-arithmetic-slices-ii-subsequence/0446-arithmetic-slices-ii-subsequence.py
class Solution:
    def numberOfArithmeticSlices(self, nums: List[int]) -> int:
        n = len(nums)
        ans = 0
        cnt = [defaultdict(int) for _ in range(n)]

        for i in range(n):
            for j in range(i):
                delta = nums[i] - nums[j]
                if delta < -2**31 or delta > 2**31 - 1:
                    continue
                diff = int(delta)
                sum_val = cnt[j][diff]
                origin = cnt[i][diff]
                cnt[i][diff] = origin + sum_val + 1
                ans += sum_val

        return int(ans)